File: Tracing/Info.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;

namespace clrays
{
    public static class Info
    {
        public static List<(string, uint)> Textures = new List<(string, uint)>();
        public static uint MetaSize, SceneSize;
        public static uint IntMapSize, FloatMapSize;
        private static Stopwatch watch = new Stopwatch();
        private static List<(string, int)> times = new List<(string, int)>();

        public static void PrintInfo()
        {
            Console.WriteLine($"Metadata: {MetaSize} B.");
            Console.WriteLine($"Scene:    {SceneSize} B.");
            Console.WriteLine($"Int FB:   {IntMapSize} B.");
            Console.WriteLine($"Float FB: {FloatMapSize} B.");
            uint sum = 0;
            for(int i = 0; i < Textures.Count; i++)
            {
                var name = Textures[i].Item1;
                var size = Textures[i].Item2;
                sum += size;
                Console.WriteLine($"Texture{i} : {name} : {size} B.");
            }
            Console.WriteLine("Totalsize: ");
            PrintSizeVerbose(sum);
            Console.WriteLine("Grand Total: ");
            sum += MetaSize + SceneSize + IntMapSize + FloatMapSize;
            PrintSizeVerbose(sum);
            int last = 0;
            for(int i = 0; i < times.Count; i++)
            {
                int elapsed = times[i].Item2 - last;
                last = times[i].Item2;
                Console.WriteLine($"{times[i].Item1}: {elapsed} ms.");
            }
            Console.WriteLine($"Total: {last}");
            watch.Stop();//why not
        }

        private static void PrintSizeVerbose(uint size)
        {
            Console.WriteLine($"    {size / Math.Pow(1024, 0)} B.");
            Console.WriteLine($"    {size / Math.Pow(1024, 1)} KB.");
            Console.WriteLine($"    {size / Math.Pow(1024, 2)} MB.");
            Console.WriteLine($"    {size / Math.Pow(1024, 3)} GB.");
        }

        public static void StartTime()
        {
            watch.Start();
        }

        public static void SetTimePoint(string name)
        {
            times.Add((name,(int)watch.ElapsedMilliseconds));
        }
    }
}

File: clrays-rs/src/info.rs
use stopwatch::{Stopwatch};

pub struct Info{
    pub textures: Vec<(String,u64)>,
    pub meta_size: u64,
    pub scene_size: u64,
    pub int_buffer_size: u64,
    pub float_buffer_size: u64,
    times: Vec<(String, i64)>,
    watch: Stopwatch,
}

impl Info{
    pub fn new() -> Self{
        Self{
            textures: Vec::new(),
            meta_size: 0,
            scene_size: 0,
            int_buffer_size: 0,
            float_buffer_size: 0,
            times: Vec::new(),
            watch: Stopwatch::new(),
        }
    }
    pub fn print_info(&self){
        println!("Metadata: {} B.", self.meta_size);
        println!("Scene:    {} B.", self.scene_size);
        println!("Int FB:   {} B.", self.int_buffer_size);
        println!("Float FB: {} B.", self.float_buffer_size);
        let mut sum = 0;
        for (i,(name,size)) in self.textures.iter().enumerate(){
            sum += size;
            println!("Texture{} : {} : {} B.", i, name, size);
        }
        println!("Totalsize: ");
        Self::print_size_verbose(sum);
        println!("Grand Total: ");
        sum += self.meta_size + self.scene_size + self.int_buffer_size + self.float_buffer_size;
        Self::print_size_verbose(sum);
        let mut last = 0;
        for (name,time) in self.times.iter(){
            let elapsed = time - last;
            last = *time;
            println!("{}: {} ms.", name, elapsed);
        }
        println!("Total: {} ms.", last);
    }

    pub fn print_size_verbose(size: u64){
        println!("      {} B.", size);
        println!("      {} KB.", size / 1024);
        println!("      {} MB.", size / 1024u64.pow(2));
        println!("      {} GB.", size / 1024u64.pow(3));
    }

    pub fn start_time(&mut self){
        self.watch.start();
    }

    pub fn set_time_point(&mut self, name: &str){
        self.times.push((name.to_string(),self.watch.elapsed_ms()));
    }
}
